luks2/backend.go

```diff
@@ -0,0 +1,147 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package luks2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/snapcore/secboot"
+	internal_luks2 "github.com/snapcore/secboot/internal/luks2"
+	"golang.org/x/sys/unix"
+)
+
+type probeDepthKeyType struct{}
+
+var (
+	probeDepthKey = probeDepthKeyType{}
+
+	devRoot              = "/dev"
+	filepathEvalSymlinks = filepath.EvalSymlinks
+	osStat               = os.Stat
+	sysfsRoot            = "/sys"
+	unixStat             = unix.Stat
+)
+
+const storageContainerBackendName = "luks2"
+
+type storageContainerBackend struct {
+	// mu is used to protect access to containers so that the
+	// Probe() method is callable safely from multiple goroutines
+	// (eg, it should be safe to call from more than 1 task in snapd).
+	mu         sync.Mutex
+	containers map[uint64]*storageContainerImpl
+}
+
+func newStorageContainerBackend() *storageContainerBackend {
+	return &storageContainerBackend{
+		containers: make(map[uint64]*storageContainerImpl),
+	}
+}
+
+// Probe implements [secboot.StorageContainerBackend.Probe].
+//
+// The path can be the path to the LUKS2 source device, or a symbolic link to it, or it can
+// be a path to an open DM volume that is backed by a LUKS2 storage device - this function
+// will walk the DM tables in this case to find the underlying source device. It identifies
+// a [StorageContainer] by its device number, and so it will always return the same instance
+// for any paths that reference the same container, regardless of what type of path is supplied.
+func (b *storageContainerBackend) Probe(ctx context.Context, path string) (secboot.StorageContainer, error) {
+	// Use the luksview package to test if the supplied path is
+	// a device or file containing a LUKS2 container.
+	_, err := newLuksView(ctx, path)
+	if err != nil {
+		if errors.Is(err, internal_luks2.ErrInvalidMagic) {
+			// We expect this error if the supplied path genuinely doesn't reference
+			// a LUKS2 container. It's possible we were supplied a path to a DM
+			// device, so walk the tree of tables to obtain a source device.
```

Comment:
Not really fan of letter the caller being confused on what path it gives and still trying to make sense of it.

What if you have a luks2 in a luks2? That does the path mean then?

Reply:
it wouldn't be too unusual too have a New* method with more strict prereqs on the arguments, and a different function that is more DWIM

Reply:
Actually, I'm wondering if `NewStorageContainer` is an appropriate name anyway, given that it's doc note contradicts that of the corresponding `StorageContainerBackend.Probe`, which says it should return the same `StorageContainer` implementation regardless of what path it's called with - it's only "new" on the first call to it. I'm going to split `Probe` into `Probe` (which assumes the path is to a source storage container) and `ProbeActivated` (which assumes the path is to a storage container that is backed by one that can be returned by `Probe` with the corresponding source path), but I think `NewStorageContainer` should probably be renamed to something else anyway.

Reply:
Ok, I've split this into `Probe` and `ProbeActivated`, and the existing `NewStorageContainer` is `FindStorageContainer` and `FindActivatedStorageContainer`.